### PR TITLE
KAZOO-5174 Fixed generation of "contact_ip" and "contact_port" fields…

### PR DESCRIPTION
**Pull request by sergey-safarov**
… in crossbar registrations

Also added support of "sips" protocol

### CODE REVIEW

**Comment by lazedo** [comment]
@sergey-safarov Hi, thanks for the pull request. can you try this instead ?

at the top
```
-include("crossbar.hrl").
-include_lib("kazoo_sip/include/kzsip_uri.hrl").
```
...

```
-spec maybe_default_port(integer(), nklib:scheme()) -> integer().
maybe_default_port(0, 'sips') -> 5061;
maybe_default_port(0, 'sip') -> 5060;
maybe_default_port(Port, _) -> Port.

-spec normalize_registration(kz_json:object()) -> kz_json:object().
normalize_registration(JObj) ->
    OriginalContact = kz_json:get_binary_value(<<"Original-Contact">>, JObj, <<>>),
    [#uri{scheme=Scheme,domain=Domain, port=Port}=_Uri] = kzsip_uri:uris(OriginalContact),
    Updaters = [fun(J) -> kz_json:delete_keys(?MASK_REG_FIELDS, J) end
               ,fun(J) ->
                        Values = [{<<"Contact-IP">>, Domain}
                                 ,{<<"Contact-Port">>, maybe_default_port(Port, Scheme)}
                                 ],
                        kz_json:set_values(Values, J)
                end
               ],
    kz_json:normalize(lists:foldr(fun(F, J) -> F(J) end, JObj, Updaters)).
```

**Comment by sergey-safarov** [comment]
@lazedo I tested new version of this functions. It is works correctly.

**Comment by lazedo** [comment]
@sergey-safarov do you want to make the changes in this PR ?

**Comment by sergey-safarov** [comment]
@lazedo I updated PR.
Also this PR not covers case when "Contact" header looks like
"sip:user@company.org;transport=tls" or "sip:user@company.org;transport=TLS"

This requirements defined in https://tools.ietf.org/html/rfc3261#section-18.1.1
 "If the port is absent, the default
   value depends on the transport.  It is 5060 for UDP, TCP and SCTP,
   5061 for TLS."

Also in section https://tools.ietf.org/html/rfc3261#section-19.1.2 of same RFC defined
  (1): The default port value is transport and scheme dependent.  The
   default  is  5060  for  sip: using UDP, TCP, or SCTP.  The default is
   5061 for sip: using TLS over TCP and sips: over TCP.

   (2): The default transport is scheme dependent.  For sip:, it is UDP.
   For sips:, it is TCP.

Please look https://tools.ietf.org/html/rfc5630#section-3.1.4 for more info.

I tried update PR to reflect this requirements but not luck. nklib is not supports arguments like "transport"

**Comment by sergey-safarov** [comment]
May be extend functionality of "kzsip_uri" module?

**Comment by lazedo** [comment]
@sergey-safarov Hi and thanks for the update. regarding the transport param you can get it using
```
 [#uri{scheme=Scheme,domain=Domain, port=Port, ext_opts=Opts}=_Uri] = kzsip_uri:uris(OriginalContact),
Transport = props:get(<<"transport">>, Opts)
...
{<<"Contact-Port">>, maybe_default_port(Port, Transport, Scheme)}
```

and change `maybe_default_port` to also get the transport and handle it with function clauses.

want to try it ?

**Comment by sergey-safarov** [comment]
Hello @lazedo 
I will try do it now

**Comment by sergey-safarov** [comment]
@lazedo I has question on code review. Pease look it.

File: applications/crossbar/src/modules/cb_registrations.erl [hunk 1]
```diff
@@ -198,36 +199,24 @@ merge_response(JObj, Regs) ->
                         end
                 end, Regs, kz_json:get_value(<<"Fields">>, JObj, [])).
 
+-spec maybe_default_port(integer(), nklib:scheme(), dict:dict()) -> integer().
+maybe_default_port(0, 'sips', _) -> 5061;
+maybe_default_port(0, 'sip', <<"TLS">>) -> 5061;
+maybe_default_port(0, 'sip', <<"tls">>) -> 5061;
+maybe_default_port(0, 'sip', _) -> 5060;
```
Comment (sergey-safarov) [inline applications/crossbar/src/modules/cb_registrations.erl]: @lazedo how strings like <<"tls">> may wrote here as 'tls'
If i write 'tls' then functions is not works. Thinks it is related to type casting. But not know what is type must be used.

File: applications/crossbar/src/modules/cb_registrations.erl [hunk 2]
```diff
@@ -198,36 +199,24 @@ merge_response(JObj, Regs) ->
                         end
                 end, Regs, kz_json:get_value(<<"Fields">>, JObj, [])).
 
+-spec maybe_default_port(integer(), nklib:scheme(), dict:dict()) -> integer().
```
Comment (lazedo) [inline applications/crossbar/src/modules/cb_registrations.erl]: `dict:dict()` should be `api_binary`

File: applications/crossbar/src/modules/cb_registrations.erl (hunk 1 above)
Comment (lazedo) [inline applications/crossbar/src/modules/cb_registrations.erl]: `<<"TLS">>` or `<<"tls">>`  is correct.
to be more dialyzer friendly you can rewrite `Transport = props:get_value` to `Transport = props:get_binary_value`

**Comment by sergey-safarov** [comment]
@lazedo PR updated and rebased on current master.
Also i test PR using sipp registration script - it works.

**Comment by lazedo** [comment]
@sergey-safarov thanks again.

**Pull request merged**